diff --git a/addons/crm/wizard/base_partner_merge.py b/addons/crm/wizard/base_partner_merge.py
index 27f3f3e..68c51a7 100644
--- a/addons/crm/wizard/base_partner_merge.py
+++ b/addons/crm/wizard/base_partner_merge.py
@@ -83,6 +83,7 @@ class MergePartnerAutomatic(models.TransientModel):
     group_by_email = fields.Boolean('Email')
     group_by_name = fields.Boolean('Name')
     group_by_is_company = fields.Boolean('Is Company')
+    group_by_customer = fields.Boolean('Customer')
     group_by_vat = fields.Boolean('VAT')
     group_by_parent_id = fields.Boolean('Parent Company')
 
@@ -359,6 +360,8 @@ class MergePartnerAutomatic(models.TransientModel):
         for field in fields:
             if field in ['email', 'name', 'vat']:
                 filters.append((field, 'IS NOT', 'NULL'))
+            filters.append(('customer', '=', True))
+            filters.append(('is_company', '=', True))
         criteria = ' AND '.join('%s %s %s' % (field, operator, value) for field, operator, value in filters)
 
         # build the query
@@ -416,8 +419,7 @@ class MergePartnerAutomatic(models.TransientModel):
             :param partner_ids : list of partner ids to sort
         """
         return self.env['res.partner'].browse(partner_ids).sorted(
-            key=lambda p: (p.active, p.create_date),
-            reverse=True,
+            key=lambda p: (p.create_uid), reverse=True
         )
 
     @api.multi
diff --git a/addons/crm/wizard/base_partner_merge_views.xml b/addons/crm/wizard/base_partner_merge_views.xml
index 6c86442..3534e40 100644
--- a/addons/crm/wizard/base_partner_merge_views.xml
+++ b/addons/crm/wizard/base_partner_merge_views.xml
@@ -34,6 +34,7 @@
                             <field name='group_by_email' />
                             <field name='group_by_name' />
                             <field name='group_by_is_company' />
+                            <field name='group_by_customer' />
                             <field name='group_by_vat' />
                             <field name='group_by_parent_id' />
                         </group>
