diff --git a/addons/sale/models/sale.py b/addons/sale/models/sale.py
index 38a4036..166f780 100644
--- a/addons/sale/models/sale.py
+++ b/addons/sale/models/sale.py
@@ -157,6 +157,9 @@ class SaleOrder(models.Model):
     procurement_group_id = fields.Many2one('procurement.group', 'Procurement Group', copy=False)
 
     product_id = fields.Many2one('product.product', related='order_line.product_id', string='Product')
+    commercial_partner_id = fields.Many2one(
+        related='partner_id.commercial_partner_id', string='Commercial Entity',
+        store=True, readonly=True)
 
     @api.model
     def _get_customer_lead(self, product_tmpl_id):
@@ -335,7 +338,9 @@ class SaleOrder(models.Model):
         invoices = {}
         references = {}
         for order in self:
-            group_key = order.id if grouped else (order.partner_invoice_id.id, order.currency_id.id)
+            group_key = order.id if grouped else (
+                order.commercial_partner_id.id,
+                                                  order.currency_id.id)
             for line in order.order_line.sorted(key=lambda l: l.qty_to_invoice < 0):
                 if float_is_zero(line.qty_to_invoice, precision_digits=precision):
                     continue
