diff --git a/addons/sale/models/sale.py b/addons/sale/models/sale.py
index 38a4036..af3d8e6 100644
--- a/addons/sale/models/sale.py
+++ b/addons/sale/models/sale.py
@@ -335,7 +335,9 @@ class SaleOrder(models.Model):
         invoices = {}
         references = {}
         for order in self:
-            group_key = order.id if grouped else (order.partner_invoice_id.id, order.currency_id.id)
+            group_key = order.id if grouped else (
+                order.commercial_partner_id.id,
+                                                  order.currency_id.id)
             for line in order.order_line.sorted(key=lambda l: l.qty_to_invoice < 0):
                 if float_is_zero(line.qty_to_invoice, precision_digits=precision):
                     continue
