diff --git a/addons/sale_stock/models/stock.py b/addons/sale_stock/models/stock.py
index 41ddb89..a3ed7f6 100644
--- a/addons/sale_stock/models/stock.py
+++ b/addons/sale_stock/models/stock.py
@@ -30,14 +30,14 @@ class StockMove(models.Model):
     @api.multi
     def assign_picking(self):
         result = super(StockMove, self).assign_picking()
-        for move in self:
-            if move.picking_id and move.picking_id.group_id:
-                picking = move.picking_id
-                order = self.env['sale.order'].sudo().search([('procurement_group_id', '=', picking.group_id.id)])
-                picking.message_post_with_view(
-                    'mail.message_origin_link',
-                    values={'self': picking, 'origin': order},
-                    subtype_id=self.env.ref('mail.mt_note').id)
+        # for move in self:
+        #     if move.picking_id and move.picking_id.group_id:
+        #         picking = move.picking_id
+        #         order = self.env['sale.order'].sudo().search([('procurement_group_id', '=', picking.group_id.id)])
+        #         picking.message_post_with_view(
+        #             'mail.message_origin_link',
+        #             values={'self': picking, 'origin': order},
+        #             subtype_id=self.env.ref('mail.mt_note').id)
         return result
 
 
